[FLINK-16653][network][tests] Introduce ResultPartitionWriterTestBase for simplifying tests

zhijiangW opened the pull request:
## What is the purpose of the change

At the moment there are at-least four implementations of `ResultPartitionWriter` interface used in unit tests. And there are about ten methods to be implemented for `ResultPartitionWriter` and most of them are dummy in tests.

When we want to extend the methods for `ResultPartitionWriter`, the above four dummy implementations in tests have to be adjusted as well, to waste a bit efforts.

Therefore the abstract ResultPartitionWriterTestBase is proposed to implement the basic dummy methods for `ResultPartitionWriter`, and the previous four instances can all extend it to only implement one or two methods based on specific requirements in tests. It will probably only need to adjust the ResultPartitionWriterTestBase when extending the `ResultPartitionWriter` interface.

## Brief change log

  - *Introduce abstract ResultPartitionWriterTestBase to implement `ResultPartitionWriter`*
  - Make previous related `ResultPartitionWriter` instances to extend `ResultPartitionWriterTestBase`

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

AHeise commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResultPartitionWriterTestBase.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * Basic behaviours of {@link ResultPartitionWriter} for test purpose.
+ */
+public abstract class ResultPartitionWriterTestBase extends TestLogger implements ResultPartitionWriter {
```
Could this class be directly be used without subclassing? Then we should remove `abstract`.

zhijiangW replied:
Good catch. My previous assumption was to only implement partial basic common methods, but after refactoring I actually implement all. So the abstract is really redundant now, i will remove it.

AHeise commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResultPartitionWriterTestBase.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * Basic behaviours of {@link ResultPartitionWriter} for test purpose.
+ */
+public abstract class ResultPartitionWriterTestBase extends TestLogger implements ResultPartitionWriter {
```
Further I propose to change the name to `MockResultPartitionWriter`. I think the suffix `TestBase` is mostly used for unit tests.

AHeise replied:
Finally, there is no benefit by implementing `TestLogger` since subclasses will not be used for implementing test cases.

zentol commented:
Why aren't we following a builder approach like we do for many other runtime components?

AHeise commented:
> Why aren't we following a builder approach like we do for many other runtime components?

Wanted to propose that, but for ResultPartitionWriter, you'd need to provide lamda hooks for half of the functions, which doesn't look like a huge win. Btw there is already a builder for ResultPartitionWriters.

zhijiangW commented:
Thanks for the review @zentol and @AHeise !

I addressed the above comments. Regarding the builder way for `ResultPartitionWriter`, we almost have the builder for specific instance class, like `ResultPartitionBuilder` for constructing `ResultPartition` instance, seldom for interface builder I guess. So I prefer the current way if no strongly advice. :)
